【Hackathon No.80】瑞芯微RK3588：通过 Paddle2ONNX 打通 5个 Paddle 模型的部署

Zheng-Bicheng opened the pull request:
第一次pr,更新picodet 在rk3588上的python 推理 + 文档

Zheng-Bicheng committed:
test

Zheng-Bicheng committed:
1. Update documentation
2. Run picodet in Python on rk3588

Zheng-Bicheng committed:
1. Fix the wrong document
2. Update new document
3. Update Python code of PP_humansegv2

Zheng-Bicheng commented:
更新PP_HumanSeg 在rk3588上的python 推理 + 文档

jiangjiajun commented:
这里整个rk3588的目录，可以移到根目录下的`hardwares`目录里面

Zheng-Bicheng commented:
> 这里整个rk3588的目录，可以移到根目录下的`hardwares`目录里面

model_zoo分支的根目录下好像没有这个文件夹，我需要新建一个吗？

jiangjiajun commented:
https://github.com/PaddlePaddle/Paddle2ONNX/tree/model_zoo/hardwares 这里

Zheng-Bicheng committed:
1. 根据给出的建议移动了picodet项目到文件夹hardwares/rk
2. 修改了项目结构，更符合原来paddle2onnx的风格
3. 修改了原始paddle2onnx部署文档的错误部分
4. 现在起，文档将统一使用markdown编写，不再上传pdf

Zheng-Bicheng committed:
1. 根据给出的建议移动了hunanseg项目到文件夹hardwares/rk
2. 修改了项目结构，更符合原来paddle2onnx的风格
3. 修改了部分文档的错误
4. 修改了代码风格

Zheng-Bicheng commented:
1. 更新代码风格
2. 修改代码目录结构
3. 根据建议移动指定的两个模型代码到指定目录

jiangjiajun commented:
谢谢你的贡献！ RKNN的部署是很多用户的需求，当前的代码可以满足一部分用户，能否同步将C++的部署代码提上来呢？

Zheng-Bicheng commented:
> 谢谢你的贡献！ RKNN的部署是很多用户的需求，当前的代码可以满足一部分用户，能否同步将C++的部署代码提上来呢？

感谢您的回复。我是想先把python的代码都部署完再写cpp的代码。cpp推理时的坑非常多，需要花费比较多的时间。先把比较简单的python代码解决，在解决 cpp。

jiangjiajun commented:
- 无论怎么设置，PP_HumanSeg的算子导出后，算子版本永远都是12。
- 这个模型在转换为rknn的时候是有bug的，无法指定输出节点，因此我们是无法自行完成模型推中结尾部分的后处理的，只能一次在npu上推理完成.
- model_zoo下的模型推理时会有很大的误差

提到的这三点问题，是通过替换模型解决吗？ 其中第2点的问题具体是指什么呢

Zheng-Bicheng commented:
> * 无论怎么设置，PP_HumanSeg的算子导出后，算子版本永远都是12。
> 
>     * 这个模型在转换为rknn的时候是有bug的，无法指定输出节点，因此我们是无法自行完成模型推中结尾部分的后处理的，只能一次在npu上推理完成.
> 
>     * model_zoo下的模型推理时会有很大的误差
> 
> 
> 提到的这三点问题，是通过替换模型解决吗？ 其中第2点的问题具体是指什么呢

关于“这个模型在转换为rknn的时候是有bug的，无法指定输出节点，因此我们是无法自行完成模型推中结尾部分的后处理的，只能一次在npu上推理完成.”和“model_zoo下的模型推理时会有很大的误差”
回答：是的，替换成了最新的PP_HumanSeg分割模型，才得以解决。但是即使替换模型后，问题3仍然出现。顺便想问一下您，有什么办法可以指定模型输出节点的shape吗？

关于“model_zoo下的模型推理时会有很大的误差”
回答：这应该是rk官方的toolkit库的问题，使用PP_HumanSeg分割模型的时候，会出现指定无论指定模型内的哪一个节点，都无法成功转换，错误代码为
`
E load_onnx: Catch exception when loading onnx model: /home/toybrick/PycharmProjects/Paddle2ONNX/hardwares/rk/segmentation/BisenetV2/weights/onnx/bisenet.onnx!
E load_onnx: Traceback (most recent call last):
E load_onnx:   File "rknn/api/rknn_base.py", line 1157, in rknn.api.rknn_base.RKNNBase.load_onnx
E load_onnx: UnboundLocalError: local variable 'inp' referenced before assignment
`

jiangjiajun commented:
我还是没明白第2点中的“一次在NPU上完成推理”指的是什么意思。

你可以试下使用onnxsimplifer处理模型，固定shape

Zheng-Bicheng commented:
> 我还是没明白第2点中的“一次在NPU上完成推理”指的是什么意思。
> 
> 你可以试下使用onnxsimplifer处理模型，固定shape

关于“我还是没明白第2点中的“一次在NPU上完成推理”指的是什么意思。”
回答：举个例子，比如模型中的某个sigmoid节点在推理时误差很大，那么我们可以把这个节点的上一个节点作为输出节点，手动完成这个节点的后续工作，这就把模型的处理分为了两步。第一步处理精度高的部分，在npu上跑；第二步，处理精度差的部分，在cpu上跑。但是无法指定输出节点，只能一步到位，全部使用npu进行推理。

jiangjiajun commented:
了解了，那现在这个误差大，是否已经找出是哪个OP的问题了，是否有联系飞桨的运营同事反馈给RKNN的研发同事来解决呢？

Zheng-Bicheng committed:
1. 更新PP-OCR_v2_det的python推理代码
2. ONNX模型固定了输入SHAPE，且修改了默认shape，效果更好
3. 重新命名了错误的文件名
4. ONNXConfig类现在能显示输入shape了
5. 更新部分错误文档

Zheng-Bicheng commented:
> 了解了，那现在这个误差大，是否已经找出是哪个OP的问题了，是否有联系飞桨的运营同事反馈给RKNN的研发同事来解决呢？

1. 没办法指定输出节点，所以我也不知道是哪个op的问题。不过误差大这个问题，在最新的PaddleDetection中的picodet模型上也有这个问题。picodet模型转换成rknn模型后是可以指定输出节点的，最后查出来是sigmoid算子的问题。所以，预计这一次也是sigmoid算子的问题，不过这也是我的猜测。
2. 还没有联系RK的研发

Zheng-Bicheng committed:
1. 更新Bisenet，rk_PC跑通，板子还有问题，在调试
2. 更新PP_OCR_v2_cls，rk_PC跑通，板子还有问题，在调试
3. 解决不能指定输出节点的bug
4. 更新部分错误文档

Zheng-Bicheng commented:
> 了解了，那现在这个误差大，是否已经找出是哪个OP的问题了，是否有联系飞桨的运营同事反馈给RKNN的研发同事来解决呢？

您好，我想问一下您，picodet有在其他npu上量化成功过的案例吗，我这边用rk的混合量化测试了一下，基本上backbone随便哪个算子float16改int8，iou就暴跌，您有出现过这种情况吗？

Zheng-Bicheng committed:
1. onnx模型使用简化过的模型，方便混合精度量化
2. 给出onnx和float16下rknn模型的精度数据和速度数据
3. 新增了picodet eval验证
4. 更新了标签，一一对应coco数据集
